### PR TITLE
[mdns] allow resolving IPv4 address for host or service with mDNSResponder

### PR DESCRIPTION
**Pull request by librasungirl**
In mdns_mdnssd.cpp, DNSServiceGetAddrInfo is called with `kDNSServiceProtocol_IPv6 | kDNSServiceProtocol_IPv4` for both service instance resolution and host subscription resolution. This indicates that either an IPv6 or an IPv4 address is explicitly expected and allowed from the mDNS responder.

However, the callback handlers `HandleGetAddrInfoResult` and `HandleResolveResult` previously discarded all non-IPv6 addresses with `VerifyOrExit(aAddress->sa_family == AF_INET6)`. This was a bug that caused hosts and services advertising only IPv4 A records to be ignored rather than resolved.

This change fixes the bug by:
- Accepting both AF_INET and AF_INET6 address families in `PublisherMDnsSd::ServiceInstanceResolution::HandleGetAddrInfoResult` and `PublisherMDnsSd::HostSubscription::HandleResolveResult`.
- Extending `Ip6Address` with `CopyFrom(const struct sockaddr &)`, `CopyFrom(const struct sockaddr_in &)` and `CopyFrom(const struct in_addr &)` to convert resolved IPv4 addresses into IPv4-mapped IPv6 addresses (`::ffff:a.b.c.d`), consistent with OpenThread's DNS-SD / discovery proxy specification.
- Adding unit test for `Ip6Address` covering IPv4-mapped addresses.

### CODE REVIEW

File: src/common/types.hpp [hunk 1]
```diff
@@ -319,6 +327,30 @@ class Ip6Address
...
+
+    /**
+     * This method copies the Ip6 address from a `sockaddr_in` structure (as an IPv4-mapped IPv6 address).
+     *
+     * @param[in] aSockAddr  The `sockaddr_in` structure to copy the Ip6 address from.
+     */
+    void CopyFrom(const struct sockaddr_in &aSockAddr);
+
+    /**
+     * This method copies the Ip6 address from a `in_addr` structure (as an IPv4-mapped IPv6 address).
```
Comment (jwhui) [inline src/common/types.hpp]:
```suggestion
     * This method copies the Ip6 address from an `in_addr` structure (as an IPv4-mapped IPv6 address).
```

File: src/mdns/mdns_mdnssd.cpp [hunk 2]
```diff
@@ -1472,17 +1472,24 @@ void PublisherMDnsSd::ServiceInstanceResolution::HandleGetAddrInfoResult(DNSServ
...
-    bool       moreComing = (aFlags & kDNSServiceFlagsMoreComing) != 0;
+    bool       isAdd;
+    bool       moreComing = false;
 
-    otbrLog(aErrorCode == kDNSServiceErr_NoError ? OTBR_LOG_INFO : OTBR_LOG_WARNING, OTBR_LOG_TAG,
-            "DNSServiceGetAddrInfo reply: flags=%" PRIu32 ", host=%s, sa_family=%u, error=%" PRId32, aFlags, aHostName,
-            static_cast<unsigned int>(aAddress->sa_family), aErrorCode);
+    if (aErrorCode != kDNSServiceErr_NoError)
+    {
+        otbrLogWarning("DNSServiceGetAddrInfo failed: error=%" PRId32, aErrorCode);
```
Comment (jwhui) [inline src/mdns/mdns_mdnssd.cpp]: Include `aHostName` in the warning log so failures can be associated with the target host.

```suggestion
        otbrLogWarning("DNSServiceGetAddrInfo failed: host=%s, error=%" PRId32, aHostName, aErrorCode);
```

File: src/mdns/mdns_mdnssd.cpp [hunk 3]
```diff
@@ -1572,17 +1579,26 @@ void PublisherMDnsSd::HostSubscription::HandleResolveResult(DNSServiceRef
...
 
-    VerifyOrExit(aErrorCode == kDNSServiceErr_NoError);
-    VerifyOrExit(aAddress->sa_family == AF_INET6);
+    VerifyOrExit(aAddress != nullptr, otbrLogWarning("DNSServiceGetAddrInfo reply has null address"));
 
-    address.CopyFrom(*reinterpret_cast<const struct sockaddr_in6 *>(aAddress));
+    otbrLogInfo("DNSServiceGetAddrInfo reply: flags=%" PRIu32 ", host=%s, sa_family=%u", aFlags, aHostName,
+                static_cast<unsigned int>(aAddress->sa_family));
+
+    SuccessOrExit(address.CopyFrom(*aAddress));
```
Comment (jwhui) [inline src/mdns/mdns_mdnssd.cpp]: If `CopyFrom()` fails, `SuccessOrExit` silently jumps to `exit`. Consider adding a warning log on failure:

```suggestion
    SuccessOrExit(address.CopyFrom(*aAddress),
                  otbrLogWarning("DNSServiceGetAddrInfo unsupported sa_family %u", static_cast<unsigned int>(aAddress->sa_family)));
```

File: tests/gtest/test_common_types.cpp [hunk 4]
```diff
@@ -26,13 +26,57 @@
...
+
 #include <gtest/gtest.h>
 
 #include "common/types.hpp"
 
 //-------------------------------------------------------------
 // Test for Ip6Address
-// TODO: Add Ip6Address tests
+
+TEST(Ip6Address, Ipv4MappedAddress)
```
Comment (jwhui) [inline tests/gtest/test_common_types.cpp:38]: How about adding test cases for `IsIp4Mapped()` on boundary addresses (e.g. ::, ::1, ::ffff:0:0) to ensure negative matches work as expected alongside the positive cases?

**Review by bukepo — APPROVED** [review]

**Review by jwhui — APPROVED** [review]
